### PR TITLE
JavaScript: Add models for popular base64 libraries.

### PR DESCRIPTION
**Pull request by xiemaisi**
[Evaluation](https://git.semmle.com/max/dist-compare-reports/blob/master/js/base64/report.md) (internal link) of the security queries on projects that contain the string "base64" shows no performance or result changes.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/Base64.qll [hunk 1]
```diff
@@ -0,0 +1,175 @@
+/**
+ * Provides classes and predicates for working with popular base64 encoders and decoders.
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/Base64.qll]: The use of "popular" seems unnecessary.

```suggestion
 * Provides classes and predicates for working with base64 encoders and decoders.
```

File: javascript/ql/src/semmle/javascript/Base64.qll [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      /** Gets the base64-encoded input passed to the decoder. */
+      abstract DataFlow::Node getInput();
+
+      /** Gets the output of the decoder. */
+      abstract DataFlow::Node getOutput();
+    }
+  }
+
+  /**
+   * A base64 decoding step, viewed as a taint-propagating data flow edge.
```
Comment (esben-semmle) [inline javascript/ql/src/semmle/javascript/Base64.qll]: Wondering (I do not have a use case in mind): do we want to consider flow through an encoder/decoder pair to be value preserving? 

Example:
```javascript
var enc = encode(x);
var dec = decode(enc);
x === dec;
```

Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/Base64.qll]: Yes, I think we do, but I also think we need flow labels to support that nicely, so I've left it out for now.

Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/Base64.qll]: It doesn't preserve object values, which I think is the main reason we care about value-preserving flow.

File: javascript/ql/src/semmle/javascript/Base64.qll [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+private class Atob extends Base64::Decode::Range, DataFlow::CallNode {
+  Atob() { this = DataFlow::globalVarRef("atob").getACall() }
+
+  override DataFlow::Node getInput() { result = getArgument(0) }
+
+  override DataFlow::Node getOutput() { result = this }
+}
+
+/** A call to `Buffer.from` with encoding `base64`. */
+private class Buffer_from extends Base64::Encode::Range, DataFlow::CallNode {
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/Base64.qll]: Shouldn't this be a decoder? It takes base64 as input. That way, creating a buffer from a user-controlled base64 string returns a tainted buffer.

File: javascript/ql/src/semmle/javascript/Base64.qll [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+  override DataFlow::Node getInput() { result = getArgument(0) }
+
+  override DataFlow::Node getOutput() { result = this }
+}
+
+/**
+ * A call to `Buffer.prototype.toString` with encoding `base64`, approximated by
+ * looking for calls to `toString` where the first argument is the string `"base64"`.
+ */
+private class Buffer_toString extends Base64::Decode::Range, DataFlow::MethodCallNode {
```
Comment (asger-semmle) [inline javascript/ql/src/semmle/javascript/Base64.qll]: Conversely, I'd say this is an encoder as it returns a base64 string.

File: javascript/ql/src/semmle/javascript/Base64.qll (hunk 3 above)
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/Base64.qll]: Ugh, yes, I think I got those the wrong way around. Thanks.

**Comment by xiemaisi** [comment]
> There are some more base64 decoders (...)

Great, thanks! I've added them all and kicked off another evaluation.

**Comment by xiemaisi** [comment]
Re-evaluation is done (report updated in place); results are the same as before.

**Pull request merged**